src/Http/Http.Extensions/src/RequestDelegateFactory.cs

```diff
@@ -222,6 +228,78 @@ public static RequestDelegateResult Create(MethodInfo methodInfo, Func<HttpConte
         return HandleRequestBodyAndCompileRequestDelegate(responseWritingMethodCall, factoryContext);
     }
 
+    private static void AddTypeProvidedMetadata(MethodInfo methodInfo, FactoryContext factoryContext)
+    {
+        const string argumentNullExceptionMessage = "The IEnumerable<object> returned from GetMetadata must not be null.";
+
+        // Get metadata from parameter types
+        var parameters = methodInfo.GetParameters();
+        foreach (var parameter in parameters)
+        {
+            if (typeof(IProvideEndpointParameterMetadata).IsAssignableFrom(parameter.ParameterType))
+            {
+                // Parameter type implements IProvideEndpointParameterMetadata
+                var metadata = GetMetadataForParameterMethod.MakeGenericMethod(parameter.ParameterType).Invoke(null, new object?[] { parameter, factoryContext.ServiceProvider });
```

Comment:
@stephentoub and @davidwrighton, this is the first example of us calling into an interface with a static abstract virtual method from late bound code. Ideally, this would be a helper like:

Reply:
> Ideally, this would be a helper like:

like... :)

Reply:
It's kinda annoying that we have to define `RequestDelegateFacotry.GetMetadataForParameter<T>(...)` in order to call `T.GetMetadata(...)` via reflection in the first place. It's defined lower down in the file, but I'll copy it here for reference.

```csharp
private static IEnumerable<object> GetMetadataForParameter<T>(ParameterInfo parameter, IServiceProvider services)
    where T : IProvideEndpointParameterMetadata
{
     return T.GetMetadata(parameter, services);
}
```

In most cases, we could just look for a public `GetMetadata` method on `T`, but what about explicit interface implementations? AFAIK, you are then forced to get the `MethodInfo` from the interface in this case. The problem is that a `static abstract` `MethodInfo` isn't invocable via reflection since there is no `this` parameter and no overload of `MethodInfo.Invoke` takes the derived type to call the static method on.

Reply:
It is technically possible to invoke the static abstract completely through reflection. We added support for static abstract methods in the `Type.GetInterfaceMap` api. However, that api is rather awkward and slow to use, so if late bound usage is critical, I would welcome proposals for updated reflection apis.